### ISSUE TITLE
Documentation on replacing window content

### ISSUE THREAD

**Issue by DapperMongoose** [cogentcore/core#1289]
### Describe the feature

I'm looking for docs on the correct way to replace the entire content of the screen in a native app.  

For example, the starting screen of my app may have connection details such as server, username, password etc.   Filling in all of the boxes and submitting would then trigger a complete re-draw in the same window with a logging in screen, and then be replaced again with the main window.

I see documentation for pages but that seems to be primarily web oriented.

Also, would one use the same method (whatever it is) if the app was also built for the web?

**Comment by kkoreilly** [comment]
You can see my comment here that answers a similar question: https://github.com/cogentcore/core/issues/1108#issuecomment-2294994762. If that does not meet your needs, please let me know. There are other possible approaches that you could consider in that case, including one based on `Maker` and one based on `DeleteChildren`.

Regardless, the same approach will work on all platforms, including the web.

**Comment by ghost** [comment]
Experimented a little with that option this evening.  It feels kinda gross to just cover up whatever the original page was, but it might be OK as long as that first screen is pretty minimal and I manage the state right before launching the new page so that boxes aren't left filled on return.  However I don't really want the back button and title bar for the new page.  Is there a way to suppress that so that I can manage the logic  to return to that page myself? (IE when the user clicks a logout button or the server kicks them out for idle/etc)

**Comment by kkoreilly** [comment]
Yes, you can disable the back button and the title bar:

```go
func main() {
	b := core.NewBody()
	bt := core.NewButton(b).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		home := core.NewBody("Home")
		core.NewText(home).SetText("Page content")
		home.NewFullDialog(bt).SetBackButton(false).SetDisplayTitle(false).Run()
	})
	b.RunMainWindow()
}
```

I also have some reservations about just covering up the original page, and there are other potential options as I mentioned. I will think some more about whether one of those other options (probably `Maker`/`Plan`) makes more sense and update you on that soon.

**Comment by kkoreilly** [comment]
This is what a basic `Maker`/`Plan` approach looks like; it is cleaner but more verbose:

```go
func main() {
	b := core.NewBody()
	page := "start"
	b.Maker(func(p *tree.Plan) {
		switch page {
		case "start":
			tree.Add(p, func(w *core.Frame) {
				bt := core.NewButton(w).SetText("Sign in")
				bt.OnClick(func(e events.Event) {
					page = "home"
					b.Update()
				})
			})
		case "home":
			tree.Add(p, func(w *core.Frame) {
				core.NewText(w).SetText("Home")
			})
		}
	})
	b.RunMainWindow()
}
```

If you like the overall structure of that better, we could add an API that handles some of the boilerplate for you.

**Comment by kkoreilly** [comment]
This is what a potential API for that could look like (not currently implemented):

```go
func main() {
	b := core.NewBody()
	pg := core.NewPages(b)
	pg.NewPage("start", func(w *core.Frame) {
		bt := core.NewButton(w).SetText("Sign in")
		bt.OnClick(func(e events.Event) {
			pg.SetPage("home")
		})
	})
	pg.NewPage("home", func(w *core.Frame) {
		core.NewText(w).SetText("Home")
	})
	b.RunMainWindow()
}
```

It is basically just a shortcut for the more verbose `Maker`/`Plan` approach I put above. I do like it, so I think I will probably implement it unless you have any feedback for how it could be structured better.

**Comment by ghost** [comment]
It seems like it's probably better, assuming my understanding of the plan machinery is correct and it's silently disposing of the old stuff.  

However it is a little hard to follow if you don't know the underlying code very well (which I'd argue is part of the point of using a framework) for instance it's not clear what w actually should be, or where\how it's getting passed to the function.

I also liked how the full window cover method would allow me to put each page in its own function.  Is there still a nice way to do that with the plan method?

My front end/GUI experience is limited so I would absolutely defer to your experience here but it seems to me like what I really want is something similar to http handlefuncs from the standard lib where you can "register" a handler and wrap it up really nicely so that main is kinda just setup and routing and ends with a listen and serve call (or in this case b.RunMainWindow() and all the real logic of each endpoint (or page in this case) gets encapsulated in a separate function.

Obviously that analogy breaks down kinda quickly because handling a request is a little different than displaying a page and swapping around between them freely, but the general idea might be compatible?

That may also be a tall order given the current architecture, but is there a way to sorta get closer to that?  Or does it not make sense/are there better options?

**Comment by kkoreilly** [comment]
Thank you for the helpful perspective. Based on my somewhat limited experience with `http.HandleFunc`, I think that this modified proposal based on the one above seems pretty similar. You can put each page in its own function, and you are basically adding "router paths" for GUI pages (you could even put slashes in the names if you wanted to). There are some issues with naming, but if you like the overall look of this proposal, then I could implement it.

```go
func main() {
	b := core.NewBody()
	pg := core.NewPages(b)
	pg.AddPage("start", start)
	pg.AddPage("home", home)
	b.RunMainWindow()
}

func start(page *core.Page) {
	bt := core.NewButton(page).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		page.SetPage("home")
	})
}

func home(page *core.Page) {
	core.NewText(page).SetText("Home")
}
```

**Comment by ghost** [comment]
This looks really promising!  How would you route from home back to start in this case?

**Comment by kkoreilly** [comment]
You could just call `SetPage` again (it would just swap the source of the content, so you could go back and forth and it would be idempotent):

```go
core.NewButton(page).SetText("Sign out").OnClick(func(e events.Event) {
	page.SetPage("start")
})
```

**Comment by ghost** [comment]
Oh OK yeah I see that now.  Because page gets passed around.  Yeah, I like that a lot!  Also allows you to conditionally route to other pages within each page's logic so you can link around at will.  There may be some gotchas that I'm not thinking of, but this sounds really slick and user friendly.

Edit: Just thought of one.  I presume that you would have the ability to pass additional parameters to the individual pages?  Thinking about state and stuff like that.

**Comment by kkoreilly** [comment]
Great! For state, you could use a closure for a page and then pass state to the actual function:

```go
func main() {
	state := ""
	b := core.NewBody()
	pg := core.NewPages(b)
	pg.NewPage("start", start)
	pg.NewPage("home", func(page *core.Page) {
		home(page, state)
	})
	b.RunMainWindow()
}

func start(page *core.Page) {
	bt := core.NewButton(page).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		page.SetPage("home")
	})
}

func home(page *core.Page, state string) {
	core.NewText(page).SetText("Home")
	core.NewButton(page).SetText("Sign out").OnClick(func(e events.Event) {
		page.SetPage("start")
	})
}
```

If you have enough things you need to manage, a struct is probably the best; here is a plausible example of how you would manage the current username:

```go
type state struct {
	user string
}

func main() {
	s := &state{}
	b := core.NewBody()
	pg := core.NewPages(b)
	pg.NewPage("start", s.start)
	pg.NewPage("home", s.home)
	b.RunMainWindow()
}

func (s *state) start(page *core.Page) {
	bt := core.NewButton(page).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		s.user = "me"
		page.SetPage("home")
	})
}

func (s *state) home(page *core.Page) {
	core.NewText(page).SetText("Welcome, " + s.user)
	core.NewButton(page).SetText("Sign out").OnClick(func(e events.Event) {
		s.user = ""
		page.SetPage("start")
	})
}
```

**Comment by ghost** [comment]
OK yeah, awesome.  Excited to give this a try! :)

**Comment by kkoreilly** [comment]
I implemented an initial version of this in #1293. If you want to look at that and see if there are any changes you think I should make, that would be great. The working [sign in tutorial](https://github.com/cogentcore/core/pull/1293/files#diff-904d44211b54b31ac7a0291be66673f245214ab4a3ba8900d3590af20551839b) demonstrates how it works. I will not merge it right now due to the name conflict with the existing `pages`, which we need to resolve, but it should be merged soon.

**Comment by ghost** [comment]
Looks pretty good to me.  Nice that it's such a small change.

**Comment by kkoreilly** [comment]
We merged the PR, so you can start using this now with the latest version of the main branch. We may make some naming changes soon, but they shouldn't have much impact and we will inform you if we do so. Please let us know if there are any issues with the widget.
